Bluetooth: tester: Support cases of feature `le sec mode 3`

lylezhu2012 opened the pull request:
* Bluetooth: tester: Support feature ISO sync receiver

Add command `BTP_GAP_BIG_CREATE_SYNC` to create BIG sync establishment.

Add event `BTP_GAP_EV_BIG_SYNC_ESTABLISHED` to notify the event that the BIG sync is established.

Add event `BTP_GAP_EV_BIS_SETUP` to notify the event that the BIS is setup.

Add event `BTP_GAP_EV_BIS_STREAM_RECEIVED` to notify the received BIS stream data.

* Bluetooth: tester: Support feature ISO broadcaster

Add command `BTP_GAP_CREATE_BIG` to create BIG.

Add command `BTP_GAP_BIS_BROADCAST` to send broadcast data.

* Bluetooth: tester: Add event of periodic BIGInfo advertising report

Add event `BTP_GAP_EV_PERIODIC_BIGINFO` to notify the received periodic BIGInfo advertising report.

* tests: bsim: tester: Add BT tester GAP ISO smoke test

Add test iso_broadcaster for GAP ISO broadcaster.
Add test iso_sync_receiver for GAP ISO sync receiver.

The changes of auto-pts is [here](https://github.com/auto-pts/auto-pts/pull/1454)

lylezhu2012 committed:
Bluetooth: tester: Support feature ISO sync receiver

Add command `BTP_GAP_BIG_CREATE_SYNC` to create BIG sync establishment.

Add event `BTP_GAP_EV_BIG_SYNC_ESTABLISHED` to notify the event that
the BIG sync is established.

Add event `BTP_GAP_EV_BIG_SYNC_LOST` to notify the event that the BIG
sync is lost.

Add event `BTP_GAP_EV_BIS_DATA_PATH_SETUP` to notify the event that
the data path of the BIS is setup.

Add event `BTP_GAP_EV_BIS_STREAM_RECEIVED` to notify the received BIS
stream data.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented:
I'll wait with reviewing this until https://github.com/auto-pts/auto-pts/pull/1454 has been more or less resolved :)

lylezhu2012 committed:
Bluetooth: tester: Add event of periodic BIGInfo advertising report

Add event `BTP_GAP_EV_PERIODIC_BIGINFO` to notify the received periodic
BIGInfo advertising report.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,376 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+
+	bt_addr_le_copy(&ev->address, &pa_sync->addr);
+	ev->bis_id = 0;
+	ev->data_len = MIN(buf->len, net_buf_simple_tailroom(iso_rx_buf));
```
Yeah, I figured, but this could cause hard-to-track test failures. 

We should probably very that the SDU size configured matches the iso_rx_buf size to avoid this happening instead. Truncating an ISO packet isn't good

lylezhu2012 replied:
Added a checking. And added a `LOG_WRN` and ignore the data in the case.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,376 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+		return;
+	}
+
+#if defined(CONFIG_BT_ISO_BROADCASTER)
```
I found there is a mistake. The macro of pre-processor is `CONFIG_BT_ISO_SYNC_RECEIVER` instead of `CONFIG_BT_ISO_BROADCASTER`.

The `info.sync_receiver` is accessed by the code block. And the field `info.sync_receiver` is controlled by the macro `CONFIG_BT_ISO_SYNC_RECEIVER`.

Thalley replied:
Regardless, we should still use IS_ENABLED over `#if defined` in this case. 

I also wonder if a `LOG_ERR` is enough if the `big` cannot be matched; should we assert instead?

lylezhu2012 replied:
Added dedicated BIG Object `iso_sync_receiver_big` and `iso_broadcaster_big` for ISO Sync Receiver and ISO Broadcaster.

The function `iso_sync_receiver_big_started_cb()` will only handle the event of `iso_sync_receiver_big`. And added `assert` if the channel of ISO Sync Receiver not found in the function.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1961,10 +1962,45 @@ static void pa_sync_recv_cb(struct bt_le_per_adv_sync *sync,
+static void pa_sync_biginfo_cb(struct bt_le_per_adv_sync *sync,
+			       const struct bt_iso_biginfo *biginfo)
+{
+	struct btp_gap_ev_periodic_biginfo_ev *ev;
```
Updated.

lylezhu2012 replied:
With the changes,
```
static void pa_sync_biginfo_cb(struct bt_le_per_adv_sync *sync,
			       const struct bt_iso_biginfo *biginfo)
{
	static struct net_buf_simple *periodic_biginfo_buf =
		NET_BUF_SIMPLE(sizeof(struct btp_gap_ev_periodic_biginfo_ev));
	struct btp_gap_ev_periodic_biginfo_ev *ev;
```

The building error is following,
```
C:/.../zephyr/tests/bluetooth/tester/src/btp_gap.c: In function 'pa_sync_biginfo_cb':
C:/.../zephyr/include/zephyr/net_buf.h:126:9: error: initializer element is not constant
  126 |         ((struct net_buf_simple *)(&(struct {        \
      |         ^
C:/.../zephyr/tests/bluetooth/tester/src/btp_gap.c:1969:17: note: in expansion of macro 'NET_BUF_SIMPLE'
 1969 |                 NET_BUF_SIMPLE(sizeof(struct btp_gap_periodic_biginfo_ev));
```

The anonymous local variable defined by `NET_BUF_SIMPLE` cannot be passed to static variable.

> Not sure what you mean by that. We do that other places: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/tester/src/btp_gatt.c#L2204

https://github.com/zephyrproject-rtos/zephyr/blob/a1455b4dfce2793d3ba5b3be33342a8ece0cf1aa/tests/bluetooth/tester/src/btp_gatt.c#L2204

The variable `buf ` is also a local variable. No such building issue found.

Thalley replied:
Ah, I guess the `static` part was the issue then?

lylezhu2012 replied:
Yes, it is.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,376 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+
+#if defined(CONFIG_BT_ISO_SYNC_RECEIVER)
+
+#if defined(CONFIG_BT_ISO_RX)
```
Updated.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,376 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+static void iso_recv(struct bt_iso_chan *chan, const struct bt_iso_recv_info *info,
+		     struct net_buf *buf)
+{
+	struct btp_gap_ev_bis_stream_received_ev *ev;
```
The data to be sent can be divided into three parts in the function `tester_event()`,
part 1, btp event header,
part 2, event header,
part 3, event payload.

From the protype of the function `void tester_event(uint8_t service, uint8_t opcode, const void *data, size_t len)`, the part1 is packed by the function `tester_event()`. Part 2 and part 3 need to be combined by the caller and pass them through parameter `data` and `len`.

If we change the protype of function `tester_event()` to `void tester_event(uint8_t service, uint8_t opcode, const void *event_header, size_t event_header_len, const void *event_payload, size_t event_payload_len)`. 
There will be two benefits:
1. No memory copy is needed,
2. Reduce the RAM usage.

Thalley replied:
Agreed. It will require ensuring that one 1 thread at a time populates the event buffer, but that should be fairly simple (and I think the function locks a mutex anyways)

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
Sure.

Thalley replied:
There aren't really that much besides the content of https://babblesim.github.io/ and https://docs.zephyrproject.org/latest/develop/test/bsim.html for setting up a local environment. 

The current GAP tests should be simple enough that you can also implement those for these new commands without needing to run locally (although verifying locally before uploading would be ideal).

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,377 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+
+	param.encryption = cp->encryption > 0 ? true : false;
+	if (param.encryption) {
+		memcpy(&param.bcode[0], &cp->broadcast_code[0], sizeof(param.bcode));
```
Updated.

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,376 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+			  BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU),
+			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static uint32_t get_next_sn(uint32_t last_sn, int64_t *last_ticks,
+			    uint32_t interval_us)
+{
+	int64_t uptime_ticks, delta_ticks;
+	uint64_t delta_us;
+	uint64_t sn_incr;
+	uint64_t next_sn;
+
+	/* Note: This does not handle wrapping of ticks when they go above
+	 * 2^(62-1)
+	 */
+	uptime_ticks = k_uptime_ticks();
+	delta_ticks = uptime_ticks - *last_ticks;
+	*last_ticks = uptime_ticks;
+
+	delta_us = k_ticks_to_us_near64((uint64_t)delta_ticks);
+	sn_incr = delta_us / interval_us;
+	next_sn = (sn_incr + last_sn);
+
+	return (uint32_t)next_sn;
+}
```
If it works, it works :D IMO the simpler and "should always work" approach is the one we use for BAP, where once the ISO channel is connected, we just start sending and then incrementing the sequence number for each SDU, which is also the approach I argued for when designing the "ISO send" BTP command, but if we stick with the current BTP design, we can continue with this approach until we see any issues

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1961,10 +1962,45 @@ static void pa_sync_recv_cb(struct bt_le_per_adv_sync *sync,
+static void pa_sync_biginfo_cb(struct bt_le_per_adv_sync *sync,
+			       const struct bt_iso_biginfo *biginfo)
+{
+	struct btp_gap_periodic_biginfo_ev *ev;
```
My goal is to reduce stack consumption. I think it is fine to use a local variable instead of global variable.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,377 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+	ev->seq_num = sys_cpu_to_le16(info->seq_num);
+	net_buf_simple_add_mem(iso_sync_receiver_buf, buf->data, ev->data_len);
+
+	tester_event(BTP_SERVICE_ID_GAP, BTP_GAP_EV_BIS_STREAM_RECEIVED,
+		     iso_sync_receiver_buf->data, iso_sync_receiver_buf->len);
```
Added the checking here https://github.com/zephyrproject-rtos/zephyr/pull/90006/files#diff-23ed7312aab28fb3100b4a3a462b51f50a4acfa4e73a7a81a7bac9f4c35359eeR2465

lylezhu2012 replied:
It is here https://github.com/zephyrproject-rtos/zephyr/pull/90006/commits/32fb5ec02f613d293523ca7dd74790d42bb717d0#diff-23ed7312aab28fb3100b4a3a462b51f50a4acfa4e73a7a81a7bac9f4c35359eeR2430.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,375 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+	const struct btp_gap_big_create_sync_cmd *cp = cmd;
+	struct bt_iso_big_sync_param param;
+
+	if ((cmd_len < sizeof(*cp)) || ((cp->encryption == 0) && (cmd_len != sizeof(*cp))) ||
+	    ((cp->encryption != 0) && (cmd_len != (sizeof(*cp) + sizeof(param.bcode))))) {
+		return BTP_STATUS_FAILED;
+	}
```
Updated the checking.
1. Firstly, verify the cmd line.
2. Then, verify the `cp->encryption` value.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,375 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+		return BTP_STATUS_FAILED;
+	}
+
+	bis_iso_qos.tx = NULL;
+	bis_iso_qos.rx = &iso_sync_receiver_qos;
+
+	bis_iso_chan.ops = &iso_sync_receiver_ops;
+
+	if (cp->num_bis > ARRAY_SIZE(bis_channels)) {
+		return BTP_STATUS_FAILED;
+	}
```
Updated.

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2385,375 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+
+	len = MIN(cp->data_len, net_buf_tailroom(buf));
+	len = MIN(len, bis_iso_chan.qos->tx->sdu);
+	net_buf_add_mem(buf, cp->data, len);
```
Updated.

Thalley reviewed (changes requested):
Looks pretty good now - a few minor outstanding comments

lylezhu2012 commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2378,400 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+	}
+
+	__ASSERT(found, "Failed to find the valid ISO info");
+
+	if (!found) {
+		LOG_ERR("Failed to find the valid ISO info");
+		return;
+	}
+
+#if defined(CONFIG_BT_ISO_SYNC_RECEIVER)
+	bt_addr_le_copy(&ev.address, &pa_sync->addr);
+	ev.latency = sys_cpu_to_le32(info.sync_receiver.latency);
+	ev.nse = info.max_subevent;
+	ev.bn = info.sync_receiver.bn;
+	ev.pto = sys_cpu_to_le32(info.sync_receiver.pto);
+	ev.irc = info.sync_receiver.irc;
+	ev.max_pdu = sys_cpu_to_le16(info.sync_receiver.max_pdu);
+	ev.iso_interval = sys_cpu_to_le16(info.iso_interval);
+
+	tester_event(BTP_SERVICE_ID_GAP, BTP_GAP_EV_BIG_SYNC_ESTABLISHED, &ev, sizeof(ev));
+#endif /* CONFIG_BT_ISO_SYNC_RECEIVER */
+}
+
+static void iso_sync_receiver_big_stopped_cb(struct bt_iso_big *big, uint8_t reason)
+{
+	struct btp_gap_big_sync_lost_ev ev;
+
+	if (big != iso_sync_receiver_big) {
+		return;
+	}
+
+	bt_addr_le_copy(&ev.address, &pa_sync->addr);
+	ev.reason = reason;
+
+	tester_event(BTP_SERVICE_ID_GAP, BTP_GAP_EV_BIG_SYNC_LOST, &ev, sizeof(ev));
+
+	iso_sync_receiver_big = NULL;
+}
+
+static struct bt_iso_big_cb iso_sync_receiver_big_cb = {
```
It is another story. The `iso_sync_receiver_big_cb` is used by calling `bt_iso_big_register_cb()`,
```
	if (IS_ENABLED(CONFIG_BT_ISO_SYNC_RECEIVER)) {
		bt_iso_big_register_cb(&iso_sync_receiver_big_cb);
	}
```

If the code block you mentioned is moved to pre-processor `#if defined(CONFIG_BT_ISO_SYNC_RECEIVER)`, the building issue will occur.

Thalley replied:
Right. 

So either it needs to be like this and have the additional `#if defined(CONFIG_BT_ISO_SYNC_RECEIVER)` when populating the `btp_gap_big_sync_established_ev`, or we need to change the 

```c
	if (IS_ENABLED(CONFIG_BT_ISO_SYNC_RECEIVER)) {
		bt_iso_big_register_cb(&iso_sync_receiver_big_cb);
	}
```

to 

```c
#if defined(CONFIG_BT_ISO_SYNC_RECEIVER)
	bt_iso_big_register_cb(&iso_sync_receiver_big_cb);
#endif /* CONFIG_BT_ISO_SYNC_RECEIVER */
```

I do prefer to use `IS_ENABLED` where applicable, but I think in this case I might actually prefer the latter, as it would clean up the `iso_sync_receiver_big_started_cb` function a little and group everything a little better IMO

Either way is OK though, so won't block this PR on this

Thalley commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -2349,6 +2378,400 @@ static uint8_t set_rpa_timeout(const void *cmd, uint16_t cmd_len, void *rsp, uin
+		break;
+	}
+
+	__ASSERT(found, "Failed to find the valid ISO info");
+
+	if (!found) {
+		LOG_ERR("Failed to find the valid ISO info");
+		return;
+	}
```
Not sure what is preferred in the BT Tester @sjanc ?

lylezhu2012 committed:
tests: bsim: tester: Add BT tester GAP ISO smoke test

Add test iso_broadcaster for GAP ISO broadcaster.
Add test iso_sync_receiver for GAP ISO sync receiver.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: tester: Support feature ISO broadcaster

Add command `BTP_GAP_CREATE_BIG` to create BIG.

Add command `BTP_GAP_BIS_BROADCAST` to send broadcast data.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: bsim: tester: Fix payload length endian issue

The payload length saved in command header should be changed from CPU
endian to little-endian.

Use `sys_cpu_to_le16()` to fix the issue.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: tester: Add LOG_* to btp.c for debug purpose

Add `LOG_ERR()` to log all reasons if status is `BTP_STATUS_FAILED`.
Add `LOG_DBG()` to log the excution result of the BTP commander.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: tester: Add debug log

Add debug log if the function `tester_gap_create_adv_instance()` is
failed.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: tester: Correct return value if ext adv create failed

The error code instead of `BTP_STATUS_FAILED` of the function
`bt_le_ext_adv_create()` should be returned.

Returns the error code returned by `bt_le_ext_adv_create()`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
Hi @Thalley, The test I added failed to run. The failed log is following,
```
d_00: @00:00:00.200196  (CPU:0): [00:00:00.200,195] <dbg> bttester: cmd_handler: cmd service 0x01 opcode 0x22 index 0x00
d_00: @00:00:00.200196  (CPU:0): [00:00:00.200,195] <err> bttester_gap: tester_gap_create_adv_instance: Failed to create ext adv(err -12)
d_00: @00:00:00.200196  (CPU:0): [00:00:00.200,195] <err> bttester_gap: tester_gap_padv_configure: Failed to create adv instance (err -12)
d_00: @00:00:00.200196  (CPU:0): 
d_00: @00:00:00.200196  (CPU:0): [00:00:00.200,195] <dbg> bttester: cmd_handler: Command returns status 1
d_00: @00:00:00.200196  (CPU:0): [00:00:00.200,195] <dbg> bttester: tester_rsp_with_index: service 0x01 opcode 0x22 index 0x00 status 0x01
d_10: @00:00:00.220215  [00:00:00.220,214] <dbg> bsim_btp: is_valid_packet_len: evt
d_10: @00:00:00.220215                                     01 00 00 01 00 01                                |......           
d_10: @00:00:00.220215  [00:00:00.220,214] <err> bsim_btp: is_valid_gap_packet_len: Unhandled opcode 0x00
d_10: @00:00:00.220215 ERROR: (WEST_TOPDIR/zephyr/tests/bsim/bluetooth/tester/src/bsim_btp.c:1876): Header len 1 does not match expected packet length for service 0x01 and opcode 0x00 (last sent opcode 0x22)
d_10: @00:00:00.220215  The TESTCASE FAILED (test return code 2)
d_11: @00:00:00.144321  TESTCASE NOT PASSED at exit (test return (1) indicates it was still in progress)
```

But in my local env, no such issue found. The error is `Failed to create ext adv(err -12)` returned by function `bt_le_ext_adv_create()`. The error code `-12` is `ENOMEM`. It means no more  advertising set can be allocated. From the setting of `overlay-bt_ll_sw_split.conf`, the value of `CONFIG_BT_EXT_ADV_MAX_ADV_SET` is 2. And in my case, the MAX advertising set is 2. So I think the issue should not occur.

And for debugging the run failure issue, I also try to add `CONFIG_BT_HCI_CORE_LOG_LEVEL_DBG=y` to `overlay-bt_ll_sw_split.conf` to try to set the log level to DBG for `HCI_CORE`. It can work in my local env. But the changes did not take any effect in PR's CI. No `LOG_DBG` message was printed in CI output.

My questions are,
- Is there any difference between running the same test in Zephyr cloud CI and running it locally? Different configuration? Additional configuration?
- How to debug the Zephyr cloud CI issue when it cannot be reproduced in local env?

lylezhu2012 commented:
> * Is there any difference between running the same test in Zephyr cloud CI and running it locally? Different configuration? Additional configuration?

I found the issue. It is another platform. The default platform is `nrf52_bsim/native`. I only test PR on it. But the issue reported on another platform `nrf5340bsim/nrf5340/cpuapp`.

Thalley reviewed (dismissed):
Nice work! :)

ardo-nordic commented on tests/bluetooth/tester/src/btp_gap.c:
```diff
@@ -1961,10 +1963,38 @@ static void pa_sync_recv_cb(struct bt_le_per_adv_sync *sync,
+{
+	struct btp_gap_periodic_biginfo_ev ev;
+
+	LOG_DBG("");
```
Remove?

lylezhu2012 commented on tests/bluetooth/tester/src/btp/btp_gap.h:
```diff
@@ -479,6 +519,63 @@ struct btp_gap_encryption_change_ev {
 	uint8_t key_size;
 } __packed;
 
+#define BTP_GAP_EV_BIG_SYNC_ESTABLISHED		0x92
```
The issue we are currently facing is that some event/command parameters are omitted when they are defined, forcing us to add V2 version to meet new requirements. 

So when defining the commands/events in this PR, some parameters are reserved even if they are not currently used.

lylezhu2012 commented:
> @lylezhu2012 please rebase and address merge conflicts

It is done.

Thalley commented:
> Can it be merged now?

Yes :) It should get merged soon

Thalley commented:
> Any reason why it is still not finished?

@kartben Is something preventing this from being merged?
